src/CloudNative.CloudEvents.AspNetCore/HttpRequestExtension.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) Cloud Native Foundation. 
+// Licensed under the Apache 2.0 license.
+// See LICENSE file in the project root for full license information.
+
+namespace CloudNative.CloudEvents
+{
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.Extensions.Primitives;
+    using Newtonsoft.Json;
+    using System;
+    using System.Net.Mime;
+
+    public static class HttpRequestExtension
+    {
+        const string HttpHeaderPrefix = "ce-";
+
+        const string SpecVersionHttpHeader1 = HttpHeaderPrefix + "cloudEventsVersion";
+
+        const string SpecVersionHttpHeader2 = HttpHeaderPrefix + "specversion";
+
+        static JsonEventFormatter jsonFormatter = new JsonEventFormatter();
+
+        /// <summary>
+        /// Converts this HTTP request into a CloudEvent object, with the given extensions.
+        /// </summary>
+        /// <param name="httpRequest">HTTP request</param>
+        /// <param name="extensions">List of extension instances</param>
+        /// <returns>A CloudEvent instance or 'null' if the request message doesn't hold a CloudEvent</returns>
+        public static CloudEvent ToCloudEvent(this HttpRequest httpRequest,
+            params ICloudEventExtension[] extensions)
+        {
+            return ToCloudEvent(httpRequest, null, extensions);
+        }
+
+        /// <summary>
+        /// Converts this HTTP request into a CloudEvent object, with the given extensions,
+        /// overriding the formatter.
+        /// </summary>
+        /// <param name="httpRequest">HTTP request</param>
+        /// <param name="formatter"></param>
+        /// <param name="extensions">List of extension instances</param>
+        /// <returns>A CloudEvent instance or 'null' if the request message doesn't hold a CloudEvent</returns>
+        public static CloudEvent ToCloudEvent(this HttpRequest httpRequest,

```

Comment:
This method shouldn't exist. We shouldn't have things that do IO but synchronous. Make it internal if you want to keep it in the formatter or remove it for now and expose the async version.

I'm also not a fan of the extension method in general but my care level is lower for that 😄. If we had to have something, I would mimic the form API:

```
public static ValueTask<CloudEvent> ReadCloudEventAsync(this HttpRequest httpRequest)
```

Reply:
@davidfowl roger, I can try and take a look. Re the extension method, are you unhappy with the method name or just the signature. I agree your name is probably better, but the library uses `ToCloudEvent()` for all the other extension method helpers that turns things into cloudevents, so I'm not sure it's worth breaking that convention

Reply:
1. Needs to be Async
2. It is worth breaking it because I'm assuming those are not doing IO (at least they shouldn't be)

Reply:
Re the second point; they're doing IO in the sense of reading from whatever stream they're given: https://github.com/cloudevents/sdk-csharp/blob/1b7f22f221ebc21c6030618ed17d1175b8c93ee8/src/CloudNative.CloudEvents/JsonEventFormatter.cs#L32-L33 
Whether that's a memory stream or something else is unknown tho.